### PR TITLE
Scale module transformers

### LINKED ISSUE #432: Add scaler transformers to Pyreal
Pyreal currently supports scaling data transforms through the `DataFrameWrapper` class, which allows us to use sklearn scalers. However, because scaling data is such a common process, pyreal code would look a lot cleaner if it directly supported these processes. 

For this PR, we will add three new Transformer classes to pyreal, in a separate transformer module called `scale`. These most likely can be implemented by just wrapping `DataFrameWrapper` instances. They should take in the same parameters as the sklearn versions.
1. `MinMaxScaler` - wraps [sklearn.preprocessing.MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler)
    - Include optional `feature_range` and `clip` parameters that get passed to wrapped transformer
2. `Normalizer` - wraps [sklearn.preprocessing.Normalizer](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.Normalizer.html#sklearn.preprocessing.Normalizer)
    - Include optional `norm` parameter that gets passed to wrapped transformer
3. `StandardScaler` - wraps [sklearn.preprocessing.StandardScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html#sklearn.preprocessing.StandardScaler)
    - Include the `with_mean` and `with_std` parameters that get passed to wrapped transformer

### PR DESCRIPTION
**Pull request by dora-x-hu**
### Closing issues

Closes #432

### Description

This PR implements sklearn scaler transformers directly into Pyreal for ease of use. Each scaler class wraps a DataFrameWrapper and calls its .wrapped_transformer attribute.

### Test Plan

Each unit tests fit(), transform(), or inverse_transform() (if applicable) for a scaler class by checking if its behavior is identical to the sklearn transformer it implements.

### CODE REVIEW

File: pyreal/transformers/scale.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
+from sklearn.preprocessing import MinMaxScaler as SklearnMinMaxScaler
```
Comment (zyteka) [inline pyreal/transformers/scale.py:1]: Pyreal no longer saves `.pkl` versions of the tutorial models - I think the ones here are a carry-over from running older code. Please remove all `.pkl` files from this PR.

File: pyreal/transformers/__init__.py [hunk 2]
```diff
@@ -59,4 +60,5 @@
     "Pandas2dToMultiIndexFrame",
     "TimeSeriesPadder",
     "BoolToIntCaster",
+    "MinMaxScaler",
```
Comment (zyteka) [inline pyreal/transformers/__init__.py:63]: Don't forget to add the other two here as well

File: pyreal/transformers/scale.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+from sklearn.preprocessing import Normalizer as SklearnNormalizer
+from sklearn.preprocessing import StandardScaler as SklearnStandardScaler
+
+from pyreal.transformers import Transformer
+from pyreal.transformers.wrappers import DataFrameWrapper
+
+
+class MinMaxScaler:
+    """
+    basically a sklearn MinMaxScaler but maintains DataFrame type
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Let's make this description a bit more formal, and make sure to add a description of what the scaler does (so people don't have to go look at sklearn's)

File: pyreal/transformers/scale.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    """
+
+    def __init__(self, feature_range=(0, 1), *, copy=True, clip=False):
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Just for consistency, please add a new line between the `"""` and the description, and capitalize the description. I would also remove implementation details from the docstring, instead talking about the actual effect (could just be something like "Initialize the transformer")

File: pyreal/transformers/scale.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        # attributes
+        self.min_ = None
+        self.scale_ = None
+        self.data_min_ = None
+        self.data_max_ = None
+        self.data_range_ = None
+        self.n_features_in_ = None
+        self.n_samples_seen_ = None
+        self.feature_names_in_ = None
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Please remove any unused class variables

File: pyreal/transformers/scale.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                Set to False to perform inplace row normalization and avoid a copy
+                (if the input is already a numpy array).
+            clip (bool, default=False):
+                Set to True to clip transformed values of held-out data
+                to provided feature range.
+        """
+        self.data_frame_wrapper = DataFrameWrapper(
+            SklearnMinMaxScaler(feature_range, copy=copy, clip=clip)
+        )
+        # self.sklearn = self.data_frame_wrapper.wrapped_transformer
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Please remove extra commented code

File: pyreal/transformers/scale.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    basically a sklearn MinMaxScaler but maintains DataFrame type
+    """
+
+    def __init__(self, feature_range=(0, 1), *, copy=True, clip=False):
+        """initialize a wrapped transformer and DataFrameWrapper, then wrap the DataFrameWrapper
+
+        Args:
+            feature_range (tuple (min, max), default=(0, 1)):
+                Desired range of transformed data.
+            copy (bool, default=True):
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: I think we can remove the copy parameter here - if we do decide to add this functionality, it will be added to the base Transformer class

File: pyreal/transformers/scale.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        self.data_min_ = None
+        self.data_max_ = None
+        self.data_range_ = None
+        self.n_features_in_ = None
+        self.n_samples_seen_ = None
+        self.feature_names_in_ = None
+
+    # methods
+
+    def fit(self, X, y=None):
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: I think you should just be able to call `self.data_frame_wrapper.fit()` to do all this functionality

File: pyreal/transformers/scale.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            feature_range (tuple (min, max), default=(0, 1)):
+                Desired range of transformed data.
+            copy (bool, default=True):
+                Set to False to perform inplace row normalization and avoid a copy
+                (if the input is already a numpy array).
+            clip (bool, default=False):
+                Set to True to clip transformed values of held-out data
+                to provided feature range.
+        """
+        self.data_frame_wrapper = DataFrameWrapper(
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Maybe a more useful name here - like `wrapped_transformer` - would be clearer

File: pyreal/transformers/scale.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+class Normalizer:
+    def __init__(self, norm="l2", *, copy=True):
+        """_summary_
+
+        Args:
+            norm (str, optional): The norm to use to normalize each non zero sample.
+                                  If norm=’max’ is used, values will be rescaled by
+                                  the maximum of the absolute values.
+                                  Can take values {‘l1’, ‘l2’, ‘max’}. Defaults to 'l2'.
+            copy (bool, optional):
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Like above, we don't need the copy parameter (#432 clarifies which parameters we want)

File: pyreal/transformers/scale.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                Set to False to perform inplace row normalization and avoid a copy
+                (if the input is already a numpy array or a scipy.sparse CSR matrix).
+                Defaults to True.
+        """
+        self.data_frame_wrapper = DataFrameWrapper(SklearnNormalizer(norm, copy=copy))
+
+    # write methods
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Please remove unused comments

File: pyreal/transformers/scale.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        self.data_max_ = None
+        self.data_range_ = None
+        self.n_features_in_ = None
+        self.n_samples_seen_ = None
+        self.feature_names_in_ = None
+
+    # methods
+
+    def fit(self, X, y=None):
+        # computes per-feature min & max (self.data_min_, self.data_max_)
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: Don't forget to add docstrings to all class functions as well

File: pyreal/transformers/scale.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+    def fit_transform(self, X, y=None, **fit_params):
+        return self.data_frame_wrapper.fit_transform(X, y, **fit_params)
+
+    def data_transform(self, X):
+        return self.data_frame_wrapper.transform(X)
+
+
+class StandardScaler(Transformer):
+    def __init__(self, *, copy=True, with_mean=True, with_std=True):
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: As above, please make the docstring style consistent and remove the copy parameter

File: pyreal/transformers/scale.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+        Args:
+            copy (bool, optional):
+                If False, try to avoid a copy and do inplace scaling instead.
+                This is not guaranteed to always work inplace;
+                e.g. if the data is not a NumPy array or scipy.sparse CSR matrix,
+                    a copy may still be returned.
+            with_mean (bool, optional):
+                If True, center the data before scaling.
+                This does not work (and will raise an exception) when attempted on sparse matrices,
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: We require inputs to be DataFrames, so you can remove mention of sparse matrices

File: pyreal/transformers/scale.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+        """
+        self.data_frame_wrapper = DataFrameWrapper(
+            SklearnStandardScaler(copy=copy, with_mean=with_mean, with_std=with_std)
+        )
+        self.mean_ = None
+        self.var_ = None
+
+    def fit(self, X, y=None, sample_weight=None):
+        ret = self.data_frame_wrapper.fit(X, y=y, sample_weight=sample_weight)
+        self.mean_ = ret.wrapped_transformer.mean_
```
Comment (zyteka) [inline pyreal/transformers/scale.py]: No need to set these member variables here, they are unused and exist in the wrapped transformer

File: tests/conftest.py [hunk 16]
```diff
@@ -197,6 +197,12 @@ def time_series_data():
     return {"np3d": np3d, "np2d": np2d, "df3d": df3d, "df2d": df2d, "nested": nested}
 
 
+@pytest.fixture()
+def minmax_scale_data():
+    x = [[2, 1, 3, 9], [4, 3, 4, 0], [6, 7, 2, 2]]
```
Comment (zyteka) [inline tests/conftest.py:203]: It is probably easier and clearer to set this to be a DataFrame here, rather than converting in each individual unit test

File: tests/transformers/test_scale.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
+from pandas import DataFrame
+from sklearn.preprocessing import MinMaxScaler as SklearnMinMaxScaler
+from sklearn.preprocessing import Normalizer as SklearnNormalizer
+from sklearn.preprocessing import StandardScaler as SklearnStandardScaler
+
+from pyreal.transformers.scale import MinMaxScaler, Normalizer, StandardScaler
+
+# MinMaxScaler tests
+data = [[2, 1, 3, 10], [4, 3, 4, 0], [6, 7, 2, 2]]
```
Comment (zyteka) [inline tests/transformers/test_scale.py]: Generally we want to avoid global variables like this in unit tests - either use the data from the conftest fixture you added (`test_fit_minmax(minmax_scale_data)`), or if it needs to be different for each unit test, set it at the beginning of the test itself

File: tests/transformers/test_scale.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+columns = ["A", "B", "C", "D"]
+
+# test whether pyreal and sklearn scalers behave the same
+
+
+def test_fit_minmax():
+    skMinMax = SklearnMinMaxScaler()
```
Comment (zyteka) [inline tests/transformers/test_scale.py]: Please be sure to follow the PEP8 style guide for naming, meaning `camel_case` names (`sk_min_max`).

File: tests/transformers/test_scale.py [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+    skMinMax.fit(data)
+
+    pdData = DataFrame(data)
+    minmax = MinMaxScaler()
+    minmax.fit(pdData)
+
+    # our scaler objects are different types; test the attributes
+    assert all(
```
Comment (zyteka) [inline tests/transformers/test_scale.py:17]: Once you remove the unused member variables, you may not be able to do these checks anymore, in which case feel free to remove them and just check for errors on fit (or check the member variables of the wrapped transformer, like `minmax.wrapped_transformer.min_`)

File: pyreal/transformers/scale.py (hunk 1 above)
Comment (zyteka) [inline pyreal/transformers/scale.py:1]: Please run all tutorial scripts one more time so the saved output does not include errors

**Review by zyteka — CHANGES_REQUESTED** [review]

File: pyreal/transformers/scale.py (hunk 1 above)
Comment (dora-x-hu) [reply pyreal/transformers/scale.py:1]: I can't find any references to .pkl files in scale.py; are you referring to files like model.pkl and transformers.pkl in the sample_applications directory?

Comment (zyteka) [reply pyreal/transformers/scale.py:1]: Yes - sorry I can't comment on the pkl files so just left this comment here, it does not have anything to do with scale.py

File: pyreal/transformers/scale.py (hunk 7 above)
Comment (dora-x-hu) [reply pyreal/transformers/scale.py]: The copy and feature_range parameters both do not exist in the base Transformer class; would they both end up in the base Transformer class? If this is the case I'm concerned about the feature_range parameter interfering with the base Transformer class (because I can't get rid of it; it doesn't have a default value).

File: pyreal/transformers/scale.py (hunk 9 above)
Comment (dora-x-hu) [reply pyreal/transformers/scale.py]: I purposely avoided using wrapped_transformer because the DataFrameWrapper class has its own wrapped_transformer attribute (so I'd end up calling self.wrapped_transformer.wrapped_transformer, which is arguably bad for understanding).

File: pyreal/transformers/scale.py (hunk 15 above)
Comment (dora-x-hu) [reply pyreal/transformers/scale.py]: To access the member variables, future users will have to do call .data_frame_wrapper.wrapped_transformer.mean_ and such; is that okay? Just checking

File: pyreal/transformers/scale.py (hunk 8 above)
Comment (dora-x-hu) [reply pyreal/transformers/scale.py]: Are you talking about initializing the attributes? fit() needs X (the dataset) as a parameter so I cannot initialize them until fit() is called or I make up a dataset for the _init_ method.

File: pyreal/transformers/scale.py (hunk 9 above)
Comment (zyteka) [reply pyreal/transformers/scale.py]: I don't think you ever need to access the DataFrameWrapper's wrapped_transformer - its functions will handle all the functionality you need.

File: pyreal/transformers/scale.py (hunk 8 above)
Comment (zyteka) [reply pyreal/transformers/scale.py]: I don't think you need any of these attributes - you never use them. Everything you need to do is handled by the wrapped DataFrameWrapper

File: pyreal/transformers/scale.py (hunk 15 above)
Comment (zyteka) [reply pyreal/transformers/scale.py]: I can't think of any reason users would need to access these member variables - if you can imagine such a scenario, we would want to write a getter (like get_mean() ), but I don't think that should be needed

**Review by zyteka — CHANGES_REQUESTED** [review]

**Review by zyteka — APPROVED** [review]
